opt_einsum/preprocessing.py

```diff
@@ -0,0 +1,244 @@
+        self.args = args
+
+    def compute(self):
+        """Compute the result of this lazy node, first computing all lazy
+        children.
+        """
+        return self.f(
+            *(
+                arg.compute() if isinstance(arg, Lazy) else arg
+                for arg in self.args
+            )
+        )
+
+    def __repr__(self):
+        return f"Lazy('{self.inds}', {self.f}, {self.args})"
+
+
+class SimplifyExpression:
+    """A callable object that transforms a list of arrays into a new simplified
+    list of arrays.
+    """
+
+    def __init__(self, lazy_inputs, lazy_outputs):
+        self._lazy_inputs = lazy_inputs
+        self._lazy_outputs = lazy_outputs
+
+    def __call__(self, arrays):
+        for a, lz in zip(arrays, self._lazy_inputs):
+            # inject arrays
+            lz.args = (a,)
+        # recursively materialize the new terms
+        return tuple(lz.compute() for lz in self._lazy_outputs)
+
+    def __repr__(self):
+        return (
+            f"SimplifyExpression('"
+            f"{','.join(lz.inds for lz in self._lazy_inputs)}->"
+            f"{','.join(lz.inds for lz in self._lazy_outputs)}')"
+        )
+
+
+def make_simplifier(*args, backend="numpy"):
+    """Take a einsum equation / specfification and return a new minimial
+    'processed' specfication and a function which will perform said processing.
+    The simplifications are as follows:
+
+        - All indices which only appear on a single input (and not the output)
+          are summed over.
+        - All indices which appear multiple times on the same term are traced.
+        - All scalars are multiplied into the smallest other term
+        - All terms with the same indices are multiplied (hamadard product
+          / elementwise) into a single term.
+
+    Parameters
+    ----------
+    args : str, or tuple
+        The einsum equation. It can already be split into inputs and output,
+        and the inputs can already be split into a list of terms.
+    backend : str, optional
+        The backend to use for any ``einsum`` operations (tracing and summing).
+
+    Returns
+    -------
+    new_spec : str, or tuple
+        The new specification, with all the simplifications applied, in the
+        same format as ``args`` was supplied.
+    simplifier : callable
+        A function which takes a list of arrays and returns the result of
+        applying the simplifications to the arrays, i.e. compatible with the
+        new einsum specification also returned.
+
+    Examples
+    --------
+
+        >>> eq = ',ab,abee,,cd,cd,dd->ac'
+        >>> arrays = helpers.build_views(eq)
+        >>> new_eq, simplifier = make_simplifier(eq)
+        >>> new_eq
+        'ab,cd,d->ac'
+
+        >>> sarrays = simplifier(arrays)
+        >>> oe.contract(new_eq, *sarrays)
+        array([[0.65245661, 0.14684493, 0.42543411, 0.32350895],
+               [0.38357005, 0.08632807, 0.25010672, 0.19018636]])
+    """
+    try:
```

Comment:
We will need quite a few tests around this tech, too bad we cannot reproduce the original parsing tech here.